tstest/natlab/natlab.go

```diff
@@ -184,27 +184,33 @@ type Machine struct {
 	interfaces []*Interface
 	routes     []routeEntry // sorted by longest prefix to shortest
 
-	conns map[netaddr.IPPort]*conn
+	conns4 map[netaddr.IPPort]*conn
+	conns6 map[netaddr.IPPort]*conn
 }
 
 func (m *Machine) deliverIncomingPacket(p []byte, dst, src netaddr.IPPort) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
-	// TODO(danderson): check behavior of dual stack sockets
-	c, ok := m.conns[dst]
-	if !ok {
-		dst = netaddr.IPPort{IP: unspecOf(dst.IP), Port: dst.Port}
-		c, ok = m.conns[dst]
+	conns := m.conns4
+	if dst.IP.Is6() {
+		conns = m.conns6
+	}
+	possibleDsts := []netaddr.IPPort{
+		dst,
+		netaddr.IPPort{IP: v6unspec, Port: dst.Port},
+		netaddr.IPPort{IP: v4unspec, Port: dst.Port},
+	}
+	for _, dst := range possibleDsts {
+		c, ok := conns[dst]
 		if !ok {
-			return
+			continue
+		}
+		select {
+		case c.in <- incomingPacket{src: src, p: p}:
+		default:
+			// Queue overflow. Just drop it.
```

Comment:
I think you want to return now, not keep looping through `possibleDsts`?

Reply:
Fixed incidentally in followup change.